[BEAM-3240] Improves development and testing instructions related to Python SDK

chamikaramj opened the pull request:
Updates contribution guide to include development and testing instructions for Python SDK.

chamikaramj commented:
R: @aaltay @davorbonaci

aaltay commented on src/contribute/contribution-guide.md:
```diff
@@ -231,10 +235,54 @@ Then you can push your local, committed changes to your (forked) repository on G
+
+For contributions to the Python code, you can use following commands to run unit tests locally and to check for lint errors. Once you send a pull request for review, these tests will be rerun using Jenkins. We recommend setting up a virtual environment before testing your code.
+
+##### Running a single test using nose
```
Should we add a section for testing with tox? (not just for lint)

chamikaramj replied:
Why is it needed ? If it simply another way to run tests I would rather avoid adding that info here.

aaltay commented on src/contribute/contribution-guide.md:
```diff
@@ -231,10 +235,54 @@ Then you can push your local, committed changes to your (forked) repository on G
 
+#### Python SDK
+
+For contributions to the Python code, you can use following commands to run unit tests locally and to check for lint errors. Once you send a pull request for review, these tests will be rerun using Jenkins. We recommend setting up a virtual environment before testing your code.
```
Maybe a TBD for people who wants to use `mvn`. FWIW `mvn clean verify` works in the python directory. It should work from the root with the right selection of project (`mvn clean verify -pl sdks/python`)

chamikaramj replied:
I think we should not add that info till https://issues.apache.org/jira/browse/BEAM-2341 and other issues around Maven+Python SDK are resolved.

aaltay replied:
I wonder if adding some information would help people coming from the mvn background. I am not sure, we can leave it as is.

chamikaramj replied:
Added instructions for running tests using Maven.

davorbonaci replied:
Don't follow -- which are "following commands"?

Fix above: Travis CI is disabled.
Overall, I'd probably generalize and say that mvn clean verify does it all, i.e., no Java SDK/Python SDK sections at all.

aaltay commented on src/contribute/contribution-guide.md:
```diff
@@ -231,10 +235,54 @@ Then you can push your local, committed changes to your (forked) repository on G
 
     $ mvn clean verify
 
+#### Python SDK
```
Should we say anything about cython? (probably not)

aaltay commented:
LGTM. Thank you. (please self merge)

chamikaramj commented:
Did following updates.

* Removed references to travis.
* Removed references to "nose" since a single test can be run without that (added this command).

Given that Maven support for Python SDK is limited (does not work from a virtual environment, cannot be used to rune a single test, etc.) I think having these instructions will be useful for potential contributors to Python SDK.

chamikaramj commented:
Regarding "I wonder should all things under ### Testing go to the testing page" I think it's better to keep instructions related to testing that applies to most PRs in the contributors guide (information I added) while keeping any additional information (for example Maven commands that users sometimes need) in a separate document. Davor, could you PTAL.

davorbonaci commented on src/contribute/contribution-guide.md:
```diff
@@ -98,6 +96,10 @@ Add your forked repository as an additional Git remote, where you’ll push your
 
 You are now ready to start developing!
 
+#### [Python SDK Only] Setup a virtual environemt
```
Setup -> Set up?

chamikaramj replied:
Done.

davorbonaci commented on src/contribute/contribution-guide.md:
```diff
@@ -98,6 +96,10 @@ Add your forked repository as an additional Git remote, where you’ll push your
 
+#### [Python SDK Only] Setup a virtual environemt
+
+We recommend setting up a virtual envioment for developing Python SDK. Please see instructions available at [Quickstart (Python)]({{ site.baseurl }}/get-started/quickstart-py/) for setting up a virtual environment.
```
at quickstart -> in quickstart?

chamikaramj replied:
Done.

davorbonaci commented on src/contribute/contribution-guide.md:
```diff
@@ -231,10 +233,31 @@ Then you can push your local, committed changes to your (forked) repository on G
 
+#### Python SDK
+
+For contributions to the Python code, you can use command given below to run unit tests locally. If you update any of the [cythonized](http://cython.org) files in python SDK, you must install "cython" package before running following command to properly test your code. We recommend setting up a virtual environment before testing your code.
```
python -> Python

chamikaramj replied:
Done.

aaltay commented:
LGTM. Thank you, please self merge.

chamikaramj commented:
Thanks.
